fenix/app/src/main/java/org/mozilla/fenix/library/history/state/HistoryUiEffectMiddleware.kt

```diff
@@ -0,0 +1,35 @@
+package org.mozilla.fenix.library.history.state
+
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.library.history.HistoryFragmentAction
+import org.mozilla.fenix.library.history.HistoryFragmentState
+
+class HistoryUiEffectMiddleware(
```

Comment:
🤔 thinking out loud here, I wonder if there's any notable slow down in navigation based on the extra processing that has to happen using middleware

Reply:
Anecdotally, in my testing with Debug Drawer, I have not seen nor felt any slowness (within previews or on-device) by having this additional overhead.

Reply:
Generally speaking, the pattern has been fast enough that there is now an [RFC](github.com/mozilla-mobile/firefox-android/pull/5353) to allow Stores to run on the main thread. There is always risk that something in the reducer chain misbehaves and slows things down (for example a middleware doing expensive computations), but that would generally be considered a solvable bug rather than a side-effect of the design principle.